README
Version: 0.2.1beta
Developer: Jeffrey Sardina 

SignRecorder is a simple program for recording and saving 
video '.avi' files for sign language data collection and
experiments. It is currently hosted on GitHub
(https://github.com/Jeffrey-Sardina/SignRecorder)
as an open-source project.

ChangeLog
v. 0.2beta to 0.2.1beta
    User-side:
        Updates how the space bar works for recording based on specifications
        Now runs reliably on MAC the way it does on windows
        Minor bug fixes and test file edits

v. 0.2alpha to 0.2beta
    User-side:
        Added a 'Show Stimuli' page in which stimuli are showed within the same window
        Many, many bug fixes that make this version semi-stable, and much more reliable than the alpha
        Program stops properly on finishing recording data
        Program quits and closes resources properly on exit (but not force-quit)
        Files chosen can been seen once they are selected in the Create Experiment page
        Support for text-stimuli removed due to lack of anticipated use
        Updated the UI to be based on text boxes, not labels
        Made UI element sizing more consistent
        Removed the UI bug in which non-current pages were still rendered
    
    Dev-side:
        Code cleaned up, with unneeded references removed
        Addition dependency PIL (Pillow) added to do rendering of images in tkinter
        Significantly better logging that is actually helpful,in finding errors
        webcam_num now defaults to 1, which makes much more sense
        Webcam searching never run, but the function remains

    Notes:
        Being a beta, this program is fairly stable but still not ready for complete use in a production environment.
        Testing has been light and mostly for non-edge cases
        Timeline for 0.2stable unknown as of yet, as it largely depends on the number of bugs found in testing the beta
        config.csv, if corrupted, should not stop the program from running

v. 0.1 to 0.2alpha
    User-side:
        Complete UI revamp. The program now runs maximized and uses a more clear dark theme
        UI is now branched into three sections--an info menu, a section to create experiments, and a section to load created experiments.
        Experiments can be created, and the result is a .csv formatted file that can be loaded at a later time to run the experiment in the recorder
        Stimuli can not be images, video, or text--but the display of those is a bit unpolished
    
    Dev-side:
        Removes a whole host of code: the Subject class and the old UI loading methods
        Moved to an OO class-based approach for the UI
        Code is now a .pyw so it does not create a terminal
        Settings class added, which links to and loads from a config file
        Logging to disk through the python logging utilities implemented
        Dependency list added (renamed to requirements so GitHub would recognize it)
        setup.py script added with dependency OpenCV-python listed
    
    Notes:
        This version is an alpha. It is quite a bit unstable, but core functionality is leaps and bounds ahead of 0.1.
            See 'Known Bugs' for more details
        Please be wary about editing the config file. There is no way to restore it yet!
        Feel free to disregard the build folder--I think I may move away from trying to compile the program. Not sure yet--that will be made clear in a later version.

v. 0.0 to 0.1
    User-side:
        Edited the About tab to contain all the data that used to be in "Credits"
        Expanded the UI view for the About tab and reduced it for the Controls tab
        Edited Controls descriptions to be more concise

    Dev-side:
        Removed old, unused code and variables
        Added a few comments
        Replaced calls to exit() with calls to sys.exit() for compatibility with older Python 3.x versions

test.py
import cv2
import tkinter as tk
from PIL import Image, ImageTk

# Load an color image
img = cv2.imread('C:/Users/jeffr/Documents/Standard Documents/Programming/Projects/VS Code DEV/SignRecorder/test_files/test_images/Blue.png')

#Rearrang the color channel
b,g,r = cv2.split(img)
img = cv2.merge((r,g,b))
im = Image.fromarray(img)

# A root window for displaying objects
root = tk.Tk()  

# Convert the Image object into a TkPhoto object
imgtk = ImageTk.PhotoImage(image=im) 

# Put it in the display window
tk.Label(root, image=imgtk, borderwidth=0, highlightthickness=0).pack() 

root.mainloop() # Start the GUI

t3.py
import tkinter
import cv2
import PIL.Image, PIL.ImageTk

class App:
    def __init__(self, window, window_title, image_path="C:/Users/jeffr/Documents/Standard Documents/Programming/Projects/VS Code DEV/SignRecorder/test_files/test_images/Blue.png"):
        self.window = window
        self.window.title(window_title)

        # Load an image using OpenCV
        self.cv_img = cv2.cvtColor(cv2.imread(image_path), cv2.COLOR_BGR2RGB)

        # Get the image dimensions (OpenCV stores image data as NumPy ndarray)
        self.height, self.width, no_channels = self.cv_img.shape

        # Create a canvas that can fit the above image
        self.canvas = tkinter.Canvas(window, width = self.width, height = self.height)
        self.canvas.pack()

        # Use PIL (Pillow) to convert the NumPy ndarray to a PhotoImage
        self.photo = PIL.ImageTk.PhotoImage(image = PIL.Image.fromarray(self.cv_img))

        # Add a PhotoImage to the Canvas
        self.canvas.create_image(0, 0, image=self.photo, anchor=tkinter.NW)

        # Button that lets the user blur the image
        self.btn_blur=tkinter.Button(window, text="Blur", width=50, command=self.blur_image)
        self.btn_blur.pack(anchor=tkinter.CENTER, expand=True)
 
        self.window.mainloop()

    # Callback for the "Blur" button
    def blur_image(self):
        self.cv_img = cv2.blur(self.cv_img, (3, 3))
        self.photo = PIL.ImageTk.PhotoImage(image = PIL.Image.fromarray(self.cv_img))
        self.canvas.create_image(0, 0, image=self.photo, anchor=tkinter.NW)


# Create a window and pass it to the Application object
App(tkinter.Tk(), "Tkinter and OpenCV")

t4.py
import cv2
import tkinter as tk
from PIL import Image, ImageTk

#Set up GUI
window = tk.Tk()  #Makes main window
window.config(background="#000000")

#Graphics window
imageFrame = tk.Frame(window, width=600, height=500)
imageFrame.grid(row=0, column=0, padx=10, pady=2)

#Capture video frames
lmain = tk.Label(imageFrame)
lmain.grid(row=0, column=0)
cap = cv2.VideoCapture('test_files/test_videos/Microsoft Edge 05_02_2019 23_54_51.mp4')

def show_frame():
    _, frame = cap.read()
    cv2image = cv2.cvtColor(frame, cv2.COLOR_BGR2RGBA)
    img = Image.fromarray(cv2image)
    imgtk = ImageTk.PhotoImage(image=img)
    lmain.imgtk = imgtk
    lmain.configure(image=imgtk)
    lmain.after(10, show_frame) 

#Slider window (slider controls stage position)
sliderFrame = tk.Frame(window, width=600, height=100)
sliderFrame.grid(row = 600, column=0, padx=10, pady=2)

show_frame()  #Display 2
window.mainloop()  #Starts GUI

SignRecorder.py
import cv2
import tkinter as tk
import threading
import sys

#recording
window = None
recording = False
just_started = True

#tk ui
record_button = None
next_button = None
controls_button = None
about_button = None
credits_button = None
data_label = None
pop_up_window = None

#exp data
subjects = []
records_since_last_next = 0
current_subject = 0

#text
controls_text = '''
Space: start / stop recording
Enter: next sign / subject
Escape: quit
'''

about_text = '''
Version: 0.1
Developer: Jeffrey Sardina 

SignRecorder is a simple program for recording and saving 
video '.avi' files for sign language data collection and
experiments. It is currently hosted on GitHub
(https://github.com/Jeffrey-Sardina/SignRecorder)
as an open-source project.
'''

def main():
    load_data()
    init_gui()
    show_gui()

def load_data():
    with open('subjects.csv', 'r') as subject_data:
        for line in subject_data:
            name, sign_string = line.split(':')
            signs = sign_string.split(',')
            subjects.append(Subject(name, signs))

def init_gui():
    global window, record_button, data_label
    #layout / theming data
    padding_x = 10
    padding_y = 5
    backcolor = '#444444'
    color = '#777777'

    #Set up GUI
    window = tk.Tk()
    window.wm_title('Sign Recorder')
    window.config(background=backcolor)

    #input
    window.bind_all('<KeyRelease>', on_key_release)

    #Label
    data_label = tk.Label(window, text = 'Press Record or Next to get started', font = (None, 20), height = 3, width = 30, background = color)
    data_label.grid(row=0, column=0, padx=padding_x, pady=padding_y)

    #Buttons
    record_button = tk.Button(window, text ="Record", command = on_button_record, font = (None, 15), height = 3, width = 30, background = color)
    record_button.grid(row=1, column=0, padx=padding_x, pady=padding_y)

    next_button = tk.Button(window, text ="Next", command = on_button_next, font = (None, 15), height = 3, width = 30, background = color)
    next_button.grid(row=2, column=0, padx=padding_x, pady=padding_y)

    blank_label = tk.Label(window, text = '', font = (None, 20), height = 1, width = 30, background = backcolor)
    blank_label.grid(row=4, column=0, padx=padding_x, pady=padding_y)

    controls_button = tk.Button(window, text ="Controls", command = on_button_controls, font = (None, 15), height = 1, width = 30, background = color)
    controls_button.grid(row=5, column=0, padx=padding_x, pady=padding_y)

    about_button = tk.Button(window, text ="About", command = on_button_about, font = (None, 15), height = 1, width = 30, background = color)
    about_button.grid(row=6, column=0, padx=padding_x, pady=padding_y)

    blank_label2 = tk.Label(window, text = '', font = (None, 20), height = 1, width = 30, background = backcolor)
    blank_label2.grid(row=7, column=0, padx=padding_x, pady=padding_y)

    exit_button = tk.Button(window, text ="Exit", command = on_button_exit, font = (None, 15), height = 3, width = 30, background = color)
    exit_button.grid(row=8, column=0, padx=padding_x, pady=padding_y)

def show_gui():
    window.mainloop()

def on_key_release(event):
    if event.keysym == 'space':
        on_button_record()
    elif event.keysym == 'Return':
        on_button_next()
    elif event.keysym == 'Escape':
        on_button_exit()

def on_button_record():
    global recording, records_since_last_next, just_started
    if just_started:
        just_started = False
        on_button_next()
    if recording:
        recording = False
        record_button.config(text = 'Record')
    else:
        recording = True
        records_since_last_next += 1
        record_button.config(text = 'Stop')
        name = subjects[current_subject].name
        sign = subjects[current_subject].current_sign_value()
        data_label.config(text = name + '; ' + sign)
        Recorder(name + ' ' + sign + '--Try' + str(records_since_last_next), 30, True).start()

def on_button_next():
    global records_since_last_next, just_started
    records_since_last_next = 0
    if just_started:
        just_started = False
    if recording:
        on_button_record()
    next_sign_or_subject()

def next_sign_or_subject():
    global current_subject, data_label
    if not current_subject >= len(subjects):
        subject = subjects[current_subject]
        name = subject.name
        sign = subject.next_sign()
    else:
        name = ''
        sign = None
        
    if sign == None:
        current_subject += 1
        if current_subject >= len(subjects):
            data_label.config(text = 'All data collected')
        else:
            name = subjects[current_subject].name
            sign = subjects[current_subject].next_sign()
            data_label.config(text = name + '; ' + sign)
    else:
        data_label.config(text = name + '; ' + sign)

def on_button_exit():
    global recording
    if recording:
        recording = False
    sys.exit()

def on_button_controls():
    global pop_up_window
    try:
        pop_up_window.destroy()
    except: #already closed by user, or never opened
        pass

    color = '#777777'

    pop_up_window = tk.Tk()
    pop_up_window.wm_title('Controls')
    pop_up_window.config(background=color)

    text = tk.Label(pop_up_window, text = controls_text, justify='left', font = (None, 20), height = 5, width = 30, background = color)
    text.grid(row=0, column=0, padx=10, pady=10)
    pop_up_window.mainloop()

def on_button_about():
    global pop_up_window
    try:
        pop_up_window.destroy()
    except: #already closed by user, or never opened
        pass

    color = '#777777'

    pop_up_window = tk.Tk()
    pop_up_window.wm_title('Controls')
    pop_up_window.config(background=color)

    text = tk.Label(pop_up_window, text = about_text, justify='left', font = (None, 20), height = 8, width = 50, background = color)
    text.grid(row=0, column=0, padx=10, pady=10)
    pop_up_window.mainloop()

class Recorder(threading.Thread):
    name = ''
    fps = 0
    mirror = False

    def __init__(self, name, fps, mirror):
        threading.Thread.__init__(self)
        self.name = name
        self.fps = fps
        self.mirror = mirror

    def run(self):
        # Capturing video from webcam:
        web_cam = cv2.VideoCapture(0)

        currentFrame = 0

        #get width and height of reading frame
        width = int(web_cam.get(cv2.CAP_PROP_FRAME_WIDTH))
        height = int(web_cam.get(cv2.CAP_PROP_FRAME_HEIGHT))
    
        # Define the codec and create VideoWriter object
        fourcc = cv2.VideoWriter_fourcc(*"XVID")
        video_writer= cv2.VideoWriter(self.name + '.avi', fourcc, self.fps, (width, height))
    
        while web_cam.isOpened():
    
            # Capture frame-by-frame
            is_reading, frame = web_cam.read()
    
            if is_reading and recording:
                if self.mirror:
                    # Mirror the if needed
                    frame = cv2.flip(frame, 1)

                # Saves for video
                video_writer.write(frame)
    
                # Display the resulting frame
                cv2.imshow('frame', frame)
            else:
                break
    
            if cv2.waitKey(1) & 0xFF == ord(' '): #quit on space
                on_button_record()
                break
    
            # To stop duplicate images
            currentFrame += 1
    
        # When everything done, release the capture
        web_cam.release()
        video_writer.release()
        cv2.destroyAllWindows()

class Subject():
    name = 'un-named'
    signs = []
    current_sign = -1

    def __init__(self, s_name, s_signs):
        self.name = s_name
        self.signs = s_signs

    def next_sign(self):
        self.current_sign += 1
        if self.current_sign < len(self.signs):
            sign = self.signs[self.current_sign].strip()
            return sign
        else:
            return None

    def current_sign_value(self):
        if self.current_sign < len(self.signs):
            return self.signs[self.current_sign].strip()
        else:
            return 'None'

main()